articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+---
+layout: default
+title: Configuring QoS from an external file
+permalink: articles/qos_file.html
+abstract:
+  This article describes how to configure QoS from an external file.
+author: '[Ivan Santiago Paunovic](https://github.com/ivanpauno)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+ROS 2 added the concept of [Quality of Service (QoS)](qos.md) settings to publishers, subscriptions, clients and services.
+As of ROS 2 Foxy Fitzroy, QoS settings can only be specified in code.
+To avoid recompiling the source code with patched QoS, node implementers have used different mechanisms to configure QoS profiles:
+- Command line arguments
+- ROS parameters
+- Combining system default QoS with vendor specific QoS profiles.
+
+Instead of having different ways of externally specifying QoS for the different entities, it would be good to have a standardized way to select QoS settings.
+
+## QoS file format
+
+DDS uses the concept of QoS profiles, in which each profile has a name.
+When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the QoS profiles file being used, and the QoS can be loaded in that way rather than from hard-coded values in the source code.
+
+If ROS would have QoS profiles names, like in DDS, collisions between those names in nodes written by different authors could happen.
+Thus, QoS profiles names would need the full power of remapping and expansion.
+
+This proposal uses the opposite approach, identifying an entity leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup> and through the use of topic/service names.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is assumed. [↩](#back_node_name_uniqueness)
+
```

Comment:
> My gut reaction is to stick with YAML, since we already use it for parameters in rcl.

That's my reaction too.

> I'm not sure there's value in having examples of multiple formats, unless there's intention to support multiple parsers.

I like proposing more than one format when I'm thinking something, I'm not sure if it adds much value or not.

> there's intention to support multiple parsers

I don't think that supporting multiple parsers is a good idea.

Reply:
I think having multiple examples is fine. I guess what I'm asking for a bit more clarity in the document. Reading through it, it gave me the impression that I should expect to write my QoS files in XML, and with the possible option of writing them in YAML. Also, the use of the work "tag" throughout makes me thing of XML (I think "tag" means something different int YAML).

Reply:
Fixed in https://github.com/ros2/design/pull/296/commits/cb65f0dbed09297743156f2adf73229e563e9f8c.